### PR TITLE
[Feature 16] adds sign-up feature

### LINKED ISSUE #16: [Feature 16] Sign-up
A user can create an account for the app

### PR DESCRIPTION
**Pull request by jenniferjinyoungyang**
What:
sign-up feature that allows a user to create an account to use the app

Why:
users should be registered so the app can share a pool of registered users for them to find their potential match.

Implementation Details

user is supposed to enter their name
user is asked to enter their name at the sign up so even though they don't update their profile, other users can see at least their name while flipping through to view profiles

there is default minimum preferred age and maximum preferred age created at the time of signup.
Even before the user updates the filters, we would like to set the minimum preferred age to 19 based on the app policy. Set the maximum preferred age to 999 although it's meaningless but at least the implementation can mirror between minimum preferred age and maximum preferred age, by not allowing only one to have the null value.

method to check if email already used is separately implemented
Although addAccount would give back userId "-1" to indicate it has failed to insert a new user if the email is already in use, thanks to the UNIQUE tag in the email column of the table account, I would like to have a separate method to check for whether the email is already used or not, so I can present more detailed information why the account creation has failed.

have a separate test database instance and clean out all database records before the test
All database records are deleted before each test so data persisted from other tests and the app won't affect the current test. To prevent the app data to be deleted due to a test, I implemented separate database instances for the test and for the app. This resolves the issue reported in https://github.com/CSC207-2023Y-UofT/course-project-tindar/issues/77

### CODE REVIEW

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java [hunk 1]
```diff
@@ -15,14 +15,18 @@
...
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Date;
 import java.util.GregorianCalendar;
 
-public class DatabaseHelper extends SQLiteOpenHelper implements EditProfileDsGateway, EditFiltersDsGateway, LoginDsGateway, LikeListDsGateway {
+public class DatabaseHelper extends SQLiteOpenHelper implements EditProfileDsGateway, EditFiltersDsGateway,
+        LoginDsGateway, LikeListDsGateway, SignUpDsGateway {
     private static DatabaseHelper dbInstance;
+    private static DatabaseHelper testDbInstance;
```
Comment (navarhontes) [inline app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java:29]: Should help solve #77?

**Review by navarhontes — DISMISSED** [review]
I am blindly approving this because I'm tired and don't want to read it, but want to have something working-ish for Milestone 4.

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java (hunk 1 above)
Comment (jenniferjinyoungyang) [reply app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java:29]: yes

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java [hunk 2]
```diff
@@ -98,6 +102,22 @@ public static synchronized DatabaseHelper getInstance(Context context) {
...
 
+    /**
+     * returns test DatabaseHelper instance. If the instance already exists it returns exiting instance, if not, it
+     * creates a instance and returns it. This insures there is only one instance of test DatabaseHelper for the
+     * application.
+     *
+     * @param context an activity's context
+     * @return an instance of the test DatabaseHelper
+     **/
+    public static synchronized DatabaseHelper getTestInstance(Context context) {
```
Comment (yiwxng) [inline app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java:113]: In terms of how this new database helper function address the issue of "potentially deleting all data in the database",  Does this function mitigate this problem by creating a new database?

**Review by yiwxng — APPROVED** [review]

File: app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java (hunk 2 above)
Comment (jenniferjinyoungyang) [reply app/src/main/java/com/courseproject/tindar/ds/DatabaseHelper.java:113]: yes this function is to get the separate instance of the test database, so it won't interfere with app database.

**Review by candidcalcifer — APPROVED** [review]

**Review by kresimirgotovac — APPROVED** [review]

**Pull request merged**